### ISSUE TITLE
项目相关

### ISSUE THREAD

**Issue by arbalest339** [arbalest339/myLeetCodeRecords#2]
- [ ] 人立方
    - [ ] 标注数据500篇，实体 个
    - [ ] 语料库4GB 实体约15万
    - [ ] 基于BERT的算法，存在问题：中美、东京 电力公司
    - [ ] 基于BERT的EL，先搜wiki，直接链接到的就链接，消歧页面转消歧。co-attention机制，S对Intro做att，再反过来，最后self-att。交叉熵损失 EL的性能 100% 95% 97.5% 实际使用时仅观测到少量假阴性预测
![image](https://user-images.githubusercontent.com/46928336/119475222-88e2e700-bd7f-11eb-8a0a-61a7cf2df57f.png)

    - [ ] 词内分词
    - [ ] 数据库设计
- [ ] 复盘核工业
    - [x] 第一阶段，算法实现，API调用，无实权，速度要求所以不用BERT
    - [x] 训练数据标注偏差大，标注员忽视了的部分关系的标注，训练时效果不错，使用时出现大量错误
    - [x] 措施：重采样，过采样，负样本加入，效果一般，其他办法（调节类型和样本的权重）
    - [ ] 通知甲方注意数据质量，但已无可挽回
    - [x] 类型匹配，限定一些实体对之间只能出现特定的关系，修改部分近似关系类型
    - [x] 第二阶段，完整系统，担任组织者，重新定义需求，完成算法的集成
    - [x] 20天速成Vue，实现事件弹幕，实体球效果
- [x] 复盘比赛
    - [x] 阿里安全+清华，辱骂文本对抗
    - [x] 没有训练数据，收集，清洗
    - [x] 构建BiLSTM目标模型，正确率不高但是能进行筛选，减少百度平台调用
    - [x] 三级词表，第一级为确定是辱骂语料中被百度开发平台判定为辱骂的词，第二级是这些句子中其他被参考模型判定为辱骂的词，最后一级是所有语料的关键词抽取
    - [x] 2，3级词表来源的语料很大程度上是本人筛选出的语料
    - [ ] 可能的改进: 用BERT构建，但意义不见得很大，因为单纯的分类只能用来进行筛选，必须要找到敏感词才能完成攻击
    - [x] 冠军队伍的思路：敏锐地察觉到噪声才是AI最大的敌人
    - [x] 研究评判标准，舍弃编辑距离，追求词向量余弦相似度，杰卡德相似度=向量每一维度上较小的数之和乘以较大的
    - [ ] 冠军改进，进一步优化编辑距离
- [x] Softmax分类器: 用交叉熵作为loss函数，用每个输出中最大的标签为结果。交叉熵=-∑pi*log(qi) q为模型概率，p为真概率
![image](https://user-images.githubusercontent.com/46928336/113735907-401f9200-972f-11eb-95a7-025fb65f9b6c.png)
- [ ] CRF
    - [x] 概率无向图模型，HHM为有向图，只能从状态i-1到i，CRF可以从任何一个状态到i
    - [ ] loss函数是发射概率+转移概率，即编码层输出真值的概率和条件随机场里真实路径的转移概率。规范化因子需要计算所有可能的路径。由真路径的概率除以所有路径概率的和，即真概率除以规范化因子。
    - [x] 改进的迭代尺度法通过改变不断优化对数似然函数改变量的下界，达到极大化对数似然函数的目的
    - [x] 动态规划维特比算法，当前步某状态的最大概率为上一步所有状态概率乘以转移到本状态的概率和 O(n*d^2)
- [ ] CRF与HMM
    - [ ] CRF 是判别模型，对问题的条件概率分布建模，而 HMM 是生成模型，对联合概率分布建模
    - [ ] CRF 用函数来表示转移概率
    - [ ] HMM用极大似然估计来作为loss，CRF则是label序列的概率与全体可能路径的比值，规范因子使得它进行了全局的归一化
    - [ ] HMM模型中存在两个假设：一是输出观察值之间严格独立，二是状态的转移过程中当前状态只与前一状态有关。
    - [ ] CRF统计了全局概率，在做归一化时考虑了数据在全局的分布，而不是仅仅在局部归一化
- [ ] LSTM
    - [ ] RNN当前步与上一步相关，无加权信息衰减
    - [ ] 序列越长前面时间步的梯度影响越少，梯度反向传播时对最前面的参数影响不大，并不擅长处理长期依赖效应
    - [ ] 如果激活函数的偏导比较小（小于1）或者为0，那么梯度随时间很容易消失
    - [ ] 遗忘门，控制上一状态对本状态的影响；输入门，控制本输入对本状态的影响；输出门，控制本状态对输出的影响
    - [ ] 当gate是关闭的，那么就会阻止对当前信息的改变，这样以前的依赖信息就会被学到。当gate是打开的时候，并不是完全替换之前的信息，而是在之前信息和现在信息之间做加权平均(候选状态是上一输出与当前输入的加权和)。所以，无论网络的深度有多深，输入序列有多长，只要gate是打开的，网络都会记住这些信息。
    - [ ] 连乘项的变化：通过控制bias比较大，使得该值接近于1
 
![image](https://user-images.githubusercontent.com/46928336/115101684-bfeffc80-9f78-11eb-94e4-ff86a136998d.png)

![image](https://user-images.githubusercontent.com/46928336/115101487-2411c100-9f77-11eb-9686-cab8f3ae77f1.png)

- [ ] GRU
    - [ ] 合并遗忘、输入门，新隐层状态是更新门乘以上一状态连接本输入，新输出状态是上一输出与本状态在输出门的线性权值(1-z)o_t-1+z*ht 
 
![image](https://user-images.githubusercontent.com/46928336/114400324-6a140100-9bd4-11eb-8ae2-c8ae6988d534.png)

- [ ] Transformer
    - [x] 6层encoder每层两个子层，第一个子层是多头自注意力，第二个是带位置的全连接层
    - [x] decoder同样6层，每一层前面再加一个带遮蔽的多头自注意力子层，遮蔽在Q*K之后，softmax之前
    - [x] Attention: Q是decoder的变换，K,V是encoder的，att = F(Q,K)*V
    - [x] self-attention: QKV分别是输入乘以一个对应的线性变换矩阵，单头自注意力层scaled-dot Product softmax(QK/(d_k)**0.5)V
    - [ ] multi-head: 多头部注意使得模型能够在不同的头共同关注来自不同表示子空间的信息。另解，将巨大的参数矩阵分为12头，减少参数随特征数的增加速度。
    - [x] 除d_k的原因：维度变大的时候点积的增长过快，可能会让softmax的梯度推到很小的区域
 ![image](https://user-images.githubusercontent.com/46928336/112810284-7153ee80-90ad-11eb-8785-384f54163d44.png)
- [x] BERT
    - [x] 模型: BERT使用的是TRM的encoder，12层768个12头注意力，[CLS]和[SEP]特殊符号
    - [x] MLM: 随机mask或替换15%的词语，预测被替换的词语，其中80%mask，10%随即替换，10%不变
    - [x] MLM的缺点：引入了不存在的[MASK], mask以piece为单位，有时会隐去完整意思的一半
    - [x] NSP：在n个下一句中选择真的下一句，一半概率为真下一句，一半概率随即替换。NSP是对真实任务有帮助的
    - [x] 2-5层表征，6-9层语法，6层以上语义
- [ ] 其他预训练语言模型
    - [ ] GPT 通过trm解码层构建且只保留了 Mask Multi-Head Attention，采用自回归机制，从左到右根据前面的预测后面的
    - [ ] Transformer XL: 引入了两点创新：循环机制（Recurrence Mechanism）获得长期依赖性和相对位置编码（Relative Positional Encoding）
        - [ ] 隐层接受上一层的输出和前面层的输出二者拼接起来做K,V
        - [ ] 计算Q*K时，K的位置编码不是绝对位置而是相对位置，是Pij而不是Pj 
![image](https://user-images.githubusercontent.com/46928336/119369131-69e44680-bce6-11eb-8fda-fc5845042ce6.png)
![image](https://user-images.githubusercontent.com/46928336/119369181-77013580-bce6-11eb-8291-5264012fe1cf.png)

    - [x] BERT-WWM(whole word mask) 用mask词语、短语级来代替mask word piece
    - [ ] ALBERT 轻量化BERT
        - [x] 分解嵌入参数，将原本的emb(vobsize*hidsize)分开为embedding矩阵(vobsize*embsize)和hidden矩阵(embsize*hidsize)
        - [ ] 在层之间共享TRM enc的参数，有助于稳定网络参数
        - [ ] 反对NSP，NSP包含了对句子话题和连贯性的预测，但是前者更容易，所以模型更倾向于用前者判断，改为句序预测
    - [ ] BART 针对文本生成（翻译）的改进
        - [ ] 训练任务：（1）5种方式加入随机噪声mask，删除（2）复原原句
        - [ ] 网络结构：BERT的编码加上GPT从左到右的解码
    - [ ] RoBerta 更大规模的BERT
        - [ ] 动态mask: 对相同的句子在不同的epoch中进行不同的mask
        - [ ] 舍弃NSP，证明了该部分几乎没有帮助，甚至有害
        - [ ] 用字节编码解决实际输入时出现未知词语的问题(将单词切做字母，然后统计字母组合的出现频率组成piece)
    - [ ] XLNet
        - [ ] 在长文本的阅读理解类的任务上提升更明显一些
        - [ ] 主流的nlp预训练模型包括两类 AR与AE，自回归的主要任务在于评估语料的概率分布，只考虑前面的词对当前词的概率。每个新单词产生后，该单词就被添加在之前生成的单词序列后面，这个序列会成为模型下一步的新输入。
        - [ ] 排列语言模型Permutation Language Model: 对句子序列完全重排列，词语的概率由每种排列中之前的词决定。在Transformer内部，通过Attention掩码，从X的输入单词里面，也就是Ti的上文和下文单词中，随机选择i-1个，放到Ti的上文位置中，把其它单词的输入通过Attention掩码隐藏掉。问题:每个词得到的依赖概率几乎一样。
        - [ ] 双流自注意力:content为正常TRM，全连接所有词，Query流隐藏单词自身以及之后的词，等于只保留位置
    - [ ] ERNIE1.0：词语级mask，短语级mask，实体级mask（知识），Dialogue Language Model 用对话代替NSP，贴吧中文数据集
    - [ ] ERNIE2.0：第一步我们通过大数据和先验知识来持续的构造无监督任务。第二步我们增量的通过 multi-task learning 来更新 ERNIE 模型。对词语、句法、语义级任务分别PT [参考](https://blog.csdn.net/c9Yv2cf9I06K2A9E/article/details/99016333)
    - [ ] K-BERT
- [ ] Torch 实现细节
    - [ ] Troch的并行化 DataParallel，net = torch.nn.DataParallel(model, device_ids=[0, 1, 2])，它将一个batch的输入划分成多个部分分配在多个设备中。双卡的时候模型本身似乎不会复制，两个卡上的数据以流水线的形式进行计算，先计算卡1的输出然后计算卡2，输出在卡1集中计算loss，loss在卡1加和，卡1计算梯度，梯度再进行反向传播，反向传播只在卡1的模型上进行。
    - [ ] Torch的底层使用 JIT (Just In Time) 即时编译器实现，JIT 包含一种名为 Torch Script 的语言，这种语言是 Python 的子语言，使用该语言可以讲torch代码序列化并支持C++ API的调用
- [ ] 解决数据不均衡的常见办法
